src/credentials/auth.ts

```diff
@@ -411,84 +489,317 @@ export class Auth implements AuthService, ConnectionManager {
+        await this.updateConnectionState(id, 'authenticating')
+
+        try {
+            const result = await callback()
+            await this.updateConnectionState(id, 'valid')
+
+            return result
+        } catch (err) {
+            await this.updateConnectionState(id, 'invalid')
+            throw err
+        }
+    }
+
     private readonly debouncedGetToken = keyedDebounce(Auth.prototype.getToken.bind(this))
     private async getToken(id: Connection['id'], provider: SsoAccessTokenProvider): Promise<SsoToken> {
         const token = await provider.getToken()
 
         return token ?? this.handleInvalidCredentials(id, () => provider.createToken())
     }
 
-    // TODO: split into 'promptInvalidCredentials' and 'authenticate' methods
+    private readonly debouncedGetCredentials = keyedDebounce(Auth.prototype.getCredentials.bind(this))
+    private async getCredentials(id: Connection['id'], provider: CredentialsProvider): Promise<Credentials> {
+        if (this.store.getProfile(id)?.metadata.connectionState !== 'valid') {
+            return this.handleInvalidCredentials(id, () => provider.getCredentials())
+        }
+
+        try {
+            return await provider.getCredentials()
+        } catch (err) {
+            return this.handleInvalidCredentials(id, () => provider.getCredentials())
+        }
+    }
+
     private async handleInvalidCredentials<T>(id: Connection['id'], refresh: () => Promise<T>): Promise<T> {
         const previousState = this.store.getProfile(id)?.metadata.connectionState
         await this.updateConnectionState(id, 'invalid')
 
         if (previousState === 'invalid') {
-            throw new ToolkitError('Credentials are invalid or expired. Try logging in again.', {
-                code: 'InvalidCredentials',
+            throw new ToolkitError('Connection is invalid or expired. Try logging in again.', {
+                code: 'InvalidConnection',
             })
         }
 
         if (previousState === 'valid') {
-            const message = localize('aws.auth.invalidCredentials', 'Credentials are expired or invalid, login again?')
+            const message = localize('aws.auth.invalidConnection', 'Connection is invalid or expired, login again?')
             const resp = await vscode.window.showInformationMessage(message, localizedText.yes, localizedText.no)
             if (resp !== localizedText.yes) {
                 throw new ToolkitError('User cancelled login', {
                     cancelled: true,
-                    code: 'InvalidCredentials',
+                    code: 'InvalidConnection',
                 })
             }
         }
 
-        const refreshed = await refresh()
-        await this.updateConnectionState(id, 'valid')
-
-        return refreshed
+        return this.authenticate(id, refresh)
     }
 
+    public readonly tryAutoConnect = once(async () => {
```

Comment:
This is a bit simpler than the impl. in `LoginManager` because we are primarily concerned with existence instead of validity of the resolved credentials.